### PR TITLE
CORDA-2148 New Cordapp plugin versioning naming conventions 

### PR DESCRIPTION
**Pull request by josecoll**
The following CorDapp jar manifest attributes:

```
["Name"]
["Implementation-Version"]
["Implementation-Vendor"]
```
are **enhanced** with distinct identifiers for Contract and WorkFlow jars:

```
["Cordapp-Contract-Name"]
["Cordapp-Contract-Version"]
["Cordapp-Contract-Vendor"]
["Cordapp-Workflow-Name"]
["Cordapp-Workflow-Version"]
["Cordapp-Workflow-Vendor"]
```

and supplemented with:

```
["Cordapp-Contract-Licence"]
["Cordapp-Workflow-Licence"]
```

An associated PR in Corda will align these new attributes for usage in Corda 4.

### CODE REVIEW

File: cordapp/src/main/kotlin/net/corda/plugins/Info.kt [hunk 1]
```diff
@@ -4,11 +4,21 @@ import org.gradle.api.tasks.Input
 
 open class Info {
     @get:Input
-    var name: String? = null
+    var cordappContractName: String? = null
     @get:Input
-    var version: String? = null
+    var cordappContractVersion: String? = null
```
Comment (szymonsztuka) [inline cordapp/src/main/kotlin/net/corda/plugins/Info.kt]: Maybe `Int` ?

File: build.gradle [hunk 2]
```diff
@@ -1,6 +1,6 @@
 buildscript {
     ext {
-        gradle_plugins_version = '4.0.37-SNAPSHOT'
+        gradle_plugins_version = '4.0.37'
```
Comment (szymonsztuka) [inline build.gradle]: This should go as separate PR, after the release.

File: cordapp/src/main/kotlin/net/corda/plugins/Info.kt (hunk 1 above)
Comment (josecoll) [reply cordapp/src/main/kotlin/net/corda/plugins/Info.kt]: @szymonsztuka lets keep it as a String in the gradle plugins to cater for future evolvability.
But we will treat it as an integer for the purposes of Contract version downgrade checking.

**Comment by chrisr3** [comment]
:man_dancing: LGTM :man_dancing:

File: cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt [hunk 3]
```diff
@@ -130,6 +141,16 @@ class CordappPlugin : Plugin<Project> {
         return Pair(targetPlatformVersion, minimumPlatformVersion)
     }
 
+    private fun parseVersion(versionStr: String?, project: Project): Int {
+        if (versionStr == null) return 1
+        return try {
+            Integer.parseInt(versionStr)
+        } catch (e: NumberFormatException) {
+            project.logger.info("Invalid version identifier: $versionStr. Defaulting to 1")
```
Comment (shamsasari) [inline cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt]: It's most likely they meant something other than 1. Better to throw an exception explaining `versionStr` is not an _integer_ version.

Comment (chrisr3) [reply cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt]: i.e. Gradle's `InvalidUserDataException`.

File: cordapp/src/test/resources/net/corda/plugins/CorDappWithContractInfo.gradle [hunk 4]
```diff
@@ -0,0 +1,25 @@
+plugins {
+    id 'java'
+    id 'net.corda.plugins.cordapp'
+}
+
+repositories {
```
Comment (chrisr3) [inline cordapp/src/test/resources/net/corda/plugins/CorDappWithContractInfo.gradle]: Perhaps refactor some of the repeated elements from these test scripts into common files? E.g. you could create a `repositories.gradle` file in `test/resources` containing:
```
repositories {
    mavenCentral()
    maven { url 'https://blah.blah.blah' }
}
```
Then replaces each `repositories` block with
```
apply from: 'repositories.gradle`
```
Finally, you'd need just need to copy this `repositories.gradle` file into the `testProjectDir`. You could do this in the `CordappTest`'s `@Before` function.

The point is that we can then update these repositories in just one place rather than `N`.

File: cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt (hunk 3 above)
Comment (josecoll) [reply cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt]: Addressed.

File: cordapp/src/test/resources/net/corda/plugins/CorDappWithContractInfo.gradle (hunk 4 above)
Comment (josecoll) [reply cordapp/src/test/resources/net/corda/plugins/CorDappWithContractInfo.gradle]: Done!

File: cordapp/src/main/kotlin/net/corda/plugins/Workflow.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package net.corda.plugins
+
+import org.gradle.api.tasks.Input
+
+open class Workflow {
+    @get:Input
```
Comment (chrisr3) [inline cordapp/src/main/kotlin/net/corda/plugins/Workflow.kt]: Are any of these parameters `@Optional`?

File: cordapp/src/test/kotlin/net/corda/plugins/Utilities.kt [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import org.junit.rules.TemporaryFolder
```
Comment (chrisr3) [inline cordapp/src/test/kotlin/net/corda/plugins/Utilities.kt]: Wot no `package`?

File: cordapp/src/test/resources/net/corda/plugins/CorDappWithInfoAll.gradle [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+apply from: 'repositories.gradle'
+
+jar {
+    baseName 'test-cordapp'
+}
+
+cordapp {
+    targetPlatformVersion target_version_arg.toInteger()
+    minimumPlatformVersion min_platform_version_arg.toInteger()
+    info {
```
Comment (shamsasari) [inline cordapp/src/test/resources/net/corda/plugins/CorDappWithInfoAll.gradle]: Surely specifying `info` alongside `contract` or `workflow` should not be allowed. Also if you specify the top level `targetPlatformVersion` or `minimumPlatformVersion` with `info`. You have to have extra checks to make sure the values are consistent.

File: cordapp/src/test/kotlin/net/corda/plugins/Utilities.kt (hunk 6 above)
Comment (josecoll) [reply cordapp/src/test/kotlin/net/corda/plugins/Utilities.kt]: bizarre how that got through.

File: cordapp/src/main/kotlin/net/corda/plugins/Workflow.kt (hunk 5 above)
Comment (josecoll) [reply cordapp/src/main/kotlin/net/corda/plugins/Workflow.kt]: There are all optional in as far as the CorDapp developer can specify either a `contract`, `workflow` or `info` `cordapp` section in their respective build.gradle files.

Comment (chrisr3) [reply cordapp/src/main/kotlin/net/corda/plugins/Workflow.kt]: Then _technically speaking_ they should also be annotated as `@Optional`. Although this probably doesn't matter unless a Gradle task uses this object as a `@Nested` field.

File: cordapp/src/test/resources/net/corda/plugins/CorDappWithInfoAll.gradle (hunk 7 above)
Comment (josecoll) [reply cordapp/src/test/resources/net/corda/plugins/CorDappWithInfoAll.gradle]: Fixed.
New top-level, `contract` and `workflow` attributes will always take precedence over older (deprecated `info` object items)

**Comment by chrisr3** [comment]
:man_cartwheeling: LGTM :man_cartwheeling:
